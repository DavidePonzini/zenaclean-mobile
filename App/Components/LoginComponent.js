import React, { Component } from 'react'
import { StyleSheet, View, Image } from 'react-native'
import { Button, Text, Input, Item, Icon } from 'native-base'
import Colors from '../Themes/Colors'
import Images from '../Themes/Images'
import api from '../Services/ApiService'
import { NavigationActions, StackActions } from 'react-navigation'

export default class LoginComponent extends Component {
  constructor (props) {
    super(props)
    this.state = {
      inputEmail: '',
      inputPassword: '',
      loginChecked: ''
    }
    api.rehydrateLogin().then(this.navigateAfterLogin)
  }

  navigateAfterLogin = (user) => {
    if (user != null) {
      this.props.navigation.dispatch(StackActions.reset({
        index: 0,
        actions: [
          NavigationActions.navigate({
            routeName: 'TabNavigator',
            params: {
              logged: true,
              inputEmail: user.email
            }
          })
        ]
      }))
    }
  }
  loginPressed = () => {
    api.logInUser(this.state.inputEmail, this.state.inputPassword, (err, user) => {
      if (err == null) {
        this.navigateAfterLogin(user)
      } else {
        console.log(err)
        this.setState({ loginChecked: 'failed' })
      }
    })
  }

  render () {
    const { navigate } = this.props.navigation
    return (
      <View style={{ ...this.props.style }}>
        <View style={styles.container}>
          <Image style={styles.logo}
            source={Images.logo}
          />
          <Item rounded style={styles.width_items}>
            <Icon name='ios-at' />
            <Input
              accessibilityLabel='login-email'
              testID={'login-email'}
              placeholder='Email'
              onChangeText={(email) => this.setState({ inputEmail: email })} />
          </Item>
          <Item rounded style={[styles.width_items, styles.input_pwd]}>
            <Icon name='ios-lock' />
            <Input
              accessibilityLabel='login-password'
              testID={'login-password'}
              placeholder='Password' secureTextEntry
              onChangeText={(password) => this.setState({ inputPassword: password })} />
          </Item>
          <Button
            accessibilityLabel='login-button'
            testID={'login-button'}
            rounded style={styles.button_login}
            onPress={() => this.loginPressed()}>
            <Text>Login</Text>
          </Button>
          {this.state.loginChecked === 'failed' && <Text style={styles.errorInputMessage}>{'Email e/o password errati'}</Text> }
          <View style={styles.fpwd_nuser}>
            <Button transparent>
              <Text accessibilityLabel='signup-button'
                testID={'signup-button'}
                style={styles.color} onPress={() => { navigate('SignUp') }}>Registrati</Text>
            </Button>
            <Button transparent
              accessibilityLabel='signup-forgot'
              testID={'signup-forgot'}>
              <Text style={styles.color}>Password Dimenticata</Text>
            </Button>
          </View>
        </View>
      </View>
    )
  }
}
const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center'
  },
  logo: {
    margin: 40,
    width: 180,
    height: 180,
    alignSelf: 'center'
  },
  button_login: {
    marginTop: 20,
    width: '90%',
    alignSelf: 'center',
    justifyContent: 'center',
    backgroundColor: Colors.accent
  },
  width_items: {
    width: '90%'
  },
  input_pwd: {
    marginTop: 20
  },
  fpwd_nuser: {
    width: '95%',
    flexDirection: 'row',
    marginTop: 20,
    justifyContent: 'space-between',
    alignItems: 'center'
  },
  color: {
    color: Colors.accent
  },
  errorInputMessage: {
    color: 'red',
    fontSize: 12
  }
})
