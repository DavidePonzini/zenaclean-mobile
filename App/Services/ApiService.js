import Secrets from 'react-native-config'

const baseUrl = Secrets.API_URL

const getAddressFromCoords = ({ lat, lng }, cb) => {
  const googleApiUrl = 'https://maps.googleapis.com/maps/api/geocode/json?address='

  fetch(googleApiUrl + lat + ',' + lng + '&key=' + Secrets.GOOGLE_MAPS_API_KEY)
    .then((response) => response.json())
    .then((responseJson) => cb(responseJson))
}

const getMarkers = (cb) => {
  fetch(baseUrl + 'markers.json')
    .then((response) => response.json())
    .then((responseJson) => { cb(responseJson) })
}

const uploadReport = (infoReport) => {
  return fetch(baseUrl + 'markers.json', {
    method: 'POST',
    body: JSON.stringify(infoReport)
  })
}
export default {
  getAddressFromCoords,
  getMarkers,
  uploadReport
}
