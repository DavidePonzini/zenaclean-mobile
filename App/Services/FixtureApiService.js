const SSNlist = ['LNGSVV60T17H509N', 'HTHHLG90R62B112N', 'BRTHZG56P68C982U']
const emailList = ['prova@gmail.com', 'prova@hotmail.it']
export const user1 = {
  email: emailList[0],
  password: '12345678',
  ssn: SSNlist[0],
  id: '5c5afd7c4be7ff1c8d7daacd'
}

const wallets = {}
wallets[user1.id] = 2
export { wallets }
export const fixtureAddress = 'Via del Camoscio'
export const fixtureMarkers = [{ address: fixtureAddress, title: 'Divano abbandonato', latitude: 44.1234, longitude: 8.1234, description: 'Divano abbandonato in via Gramsci', timestamp: '2018-12-09T14:40:35 -02:00', approved_positive: false, approved_negative: false, _id: '5c6580a33a7f647afccd8920', user_id: user1.id, url: 'images/667856780e39f0bbe7407bfdae603f8e', voted_positive: false, voted_negative: false }]
export const fixtureMessageOk = 'Registrazione effettuata con successo!'
export const fixtureMessageNotOk = 'Utente giÃ  esistente'
export const fixtureUserId = 'verylonguserid'
const totVoti = [32, 45, 21]
export let userId = null

const logInUser = (email, password, cb) => {
  return new Promise(resolve => {
    (user1.email === email && user1.password === password) ? resolve({ status: 'ok' }) : resolve({ status: 'no' })
  }).then((res) => {
    if (res.status === 'ok') {
      userId = fixtureUserId
      cb(null, { id: userId, email, eth_address: null })
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const getAddressFromCoords = ({ lat, lng }, cb) => {
  return new Promise((resolve) => {
    resolve(fixtureAddress)
  }).then((res) => { cb(null, res) })
}

const getMarkers = (neLat, swLat, swLng, neLng, cb) => {
  return new Promise((resolve) => {
    resolve(JSON.stringify(fixtureMarkers))
  }).then((res) => { cb(JSON.parse(res)) })
}

const uploadReport = (infoReport, cb) => {
  return new Promise((resolve) => {
    resolve({ status: 'ok' })
  }).then(() => {
    cb(null)
  })
}

const registerUser = (email, ssn, password, cb) => {
  return new Promise((resolve) => {
    let checked = false
    for (let i = 0; i < SSNlist.length; i++) {
      if (SSNlist[i] === ssn) checked = true
    }
    for (let i = 0; i < emailList.length; i++) {
      if (emailList[i] === email) checked = true
    }
    if (!checked) { resolve(fixtureMessageOk) }
    if (checked) { resolve(fixtureMessageNotOk) }
  }).then((res) => { cb(null, res) })
}

const voteReport = (v, report, cb) => {
  return new Promise((resolve) => {
    let old = totVoti[report]
    let current = totVoti[report] + 1
    if (old < current) resolve({ status: 'ok' })
    else resolve({ status: 'no' })
  }).then((res, current) => {
    if (res.status === 'ok') {
      totVoti[report] = current
      cb(null)
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const changePassword = (email, oldPassword, newPassword, confirmPassword, cb) => {
  return new Promise((resolve) => {
    if (user1.password === oldPassword && newPassword === confirmPassword) resolve({ status: 'ok' })
    else resolve({ status: 'no' })
  }).then((res) => {
    if (res.status === 'ok') {
      user1.password = newPassword
      cb(null)
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const isLoggedIn = () => userId != null

const logoutUser = () => {
  userId = null
}
const rehydrateLogin = () => {
  return new Promise((resolve, reject) => {
    if (userId != null) {
      resolve(userId)
    } else {
      reject(new Error('Not logged in'))
    }
  })
}

const getBalance = (eth_addr, cb) => {
  return new Promise(resolve => {
    resolve({ value: wallets[user1.id] })
  }).then(res => {
    cb(res.value)
  })
}

export default {
  getAddressFromCoords,
  getMarkers,
  uploadReport,
  registerUser,
  logInUser,
  changePassword,
  voteReport,
  isLoggedIn,
  logoutUser,
  rehydrateLogin,
  getBalance
}
