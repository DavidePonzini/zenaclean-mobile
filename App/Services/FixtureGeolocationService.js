const fixturePosition = {
  coords: {
    accuracy: 1135,
    altitude: null,
    altitudeAccuracy: null,
    heading: null,
    latitude: 44.4135423,
    longitude: 8.9554223,
    speed: null
  },
  timestamp: 1547641567823
}

const fixtureRegion = {
  latitude: 44.4135423,
  longitude: 8.9554223,
  latitudeDelta: 0.1,
  longitudeDelta: 0.1
}

let lastLocation = fixturePosition
let currentRegion = fixtureRegion

const setCurrentRegion = reg => { currentRegion = reg }
const getCurrentRegion = () => currentRegion

const geolocateOnce = async () => {
  return new Promise(resolve => {
    let pos = fixturePosition
    lastLocation = pos
    resolve(pos)
  })
}

const getLatestPosition = () => lastLocation

const geolocateContinuously = (cb, cbErr, options) => {
  return setInterval(cb, 100)
}

const stopGeolocating = (id) => {
  clearInterval(id)
}

export default {
  geolocateOnce,
  geolocateContinuously,
  stopGeolocating,
  getLatestPosition,
  getCurrentRegion,
  setCurrentRegion
}
