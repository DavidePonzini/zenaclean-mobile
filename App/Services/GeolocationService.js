let lastLocation = null

// initial region
let currentRegion = {
  latitude: 44.4056,
  longitude: 8.9463,
  latitudeDelta: 0.1,
  longitudeDelta: 0.1
}

const setCurrentRegion = reg => { currentRegion = reg }
const getCurrentRegion = () => currentRegion

const geolocateOnce = async () => {
  return new Promise(resolve => {
    navigator.geolocation.getCurrentPosition((pos) => {
      lastLocation = pos
      resolve(pos)
    })
  })
}

const geolocateContinuously = (cb, cbErr, options) => {
  return navigator.geolocation.watchPosition((pos) => {
    cb(pos)
    lastLocation = pos
  }, cbErr, options)
}

const stopGeolocating = (id) => {
  navigator.geolocation.clearWatch(id)
}

const getLatestPosition = () => lastLocation

export default {
  geolocateOnce,
  geolocateContinuously,
  stopGeolocating,
  getLatestPosition,
  getCurrentRegion,
  setCurrentRegion
}
