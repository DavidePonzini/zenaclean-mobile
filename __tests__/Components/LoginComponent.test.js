import React from 'react'
import LoginComponent from '../../App/Components/LoginComponent'

import { user1 } from '../../App/Services/FixtureApiService'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import 'isomorphic-fetch'
import { shallowToJson } from 'enzyme-to-json'

configure({ adapter: new Adapter() })

jest.mock('../../App/Services/ApiService', () => require('../../App/Services/FixtureApiService'))
const navigation = { navigate: jest.fn(), dispatch: jest.fn() }

describe('LoginComponent tests', () => {
  const wrapper = shallow(<LoginComponent navigation={navigation} />)
  it('renders correctly', () => {
    wrapper.update()
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('navigates when successful login', async (done) => {
    await logInCorrectly(wrapper)
    await expect(navigation.dispatch).toHaveBeenCalled()
    done()
  })

  it('navigates when it finds a persisted session', async (done) => {
    const wrapper = shallow(<LoginComponent navigation={navigation} />)
    await logInCorrectly(wrapper)
    wrapper.unmount()
    shallow(<LoginComponent navigation={navigation} />)
    await expect(navigation.dispatch).toHaveBeenCalled()
    done()
  })

  it('shows a popup on failed login attempt', async (done) => {
    const emailInput = wrapper.find({ testID: 'login-email' }).first()
    emailInput.simulate('changeText', 'pippo')
    const pwdInput = wrapper.find({ testID: 'login-password' }).first()
    pwdInput.simulate('changeText', 'pippo')
    wrapper.update()
    const loginButton = wrapper.find({ testID: 'login-button' }).first()
    await loginButton.simulate('Press')
    await expect(wrapper.find({ testID: 'login-failed' })).toHaveLength(1)
    done()
  })
})

const logInCorrectly = async (wrapper) => {
  const emailInput = wrapper.find({ testID: 'login-email' }).first()
  emailInput.simulate('changeText', user1.email)
  const pwdInput = wrapper.find({ testID: 'login-password' }).first()
  pwdInput.simulate('changeText', user1.password)
  wrapper.update()
  const loginButton = wrapper.find({ testID: 'login-button' }).first()
  await loginButton.simulate('Press')
}