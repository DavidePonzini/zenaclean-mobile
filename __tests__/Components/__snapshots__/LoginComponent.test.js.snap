// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent tests renders correctly 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <Component
      source={
        Object {
          "testUri": "../../../App/Assets/Images/logo.png",
        }
      }
      style={
        Object {
          "alignSelf": "center",
          "height": 180,
          "margin": 40,
          "width": 180,
        }
      }
    />
    <Styled(Item)
      rounded={true}
      style={
        Object {
          "width": "90%",
        }
      }
    >
      <Styled(Icon)
        name="ios-at"
      />
      <Styled(Input)
        accessibilityLabel="login-email"
        onChangeText={[Function]}
        placeholder="Email"
        testID="login-email"
      />
    </Styled(Item)>
    <Styled(Item)
      rounded={true}
      style={
        Array [
          Object {
            "width": "90%",
          },
          Object {
            "marginTop": 20,
          },
        ]
      }
    >
      <Styled(Icon)
        name="ios-lock"
      />
      <Styled(Input)
        accessibilityLabel="login-password"
        onChangeText={[Function]}
        placeholder="Password"
        secureTextEntry={true}
        testID="login-password"
      />
    </Styled(Item)>
    <Styled(Button)
      accessibilityLabel="login-button"
      onPress={[Function]}
      rounded={true}
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "dodgerblue",
          "justifyContent": "center",
          "marginTop": 20,
          "width": "90%",
        }
      }
      testID="login-button"
    >
      <Styled(Text)>
        Login
      </Styled(Text)>
    </Styled(Button)>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 20,
          "width": "95%",
        }
      }
    >
      <Styled(Button)
        transparent={true}
      >
        <Styled(Text)
          accessibilityLabel="signup-button"
          onPress={[Function]}
          style={
            Object {
              "color": "dodgerblue",
            }
          }
          testID="signup-button"
        >
          Registrati
        </Styled(Text)>
      </Styled(Button)>
      <Styled(Button)
        accessibilityLabel="signup-forgot"
        testID="signup-forgot"
        transparent={true}
      >
        <Styled(Text)
          style={
            Object {
              "color": "dodgerblue",
            }
          }
        >
          Password Dimenticata
        </Styled(Text)>
      </Styled(Button)>
    </View>
  </View>
</View>
`;
